site/docs/v1/tech/guides/advanced-registration-forms.adoc

```diff
@@ -0,0 +1,540 @@
+```
+
+The `convert.py` script turns embedded newlines into real ones:
+
+```python
+import sys
+
+OUTPUT = sys.stdin.read()
+formatted_output = OUTPUT.replace('\\n', '\n')
+print(formatted_output)
+```
+
+Running this script after updating the API key and theme ID will create a `defaultmessages.txt` file in the current directory. This script downloads only the messages file, but could be extended to retrieve other theme attributes. The `defaultmessages.txt` file contents look like this:
+
+```
+#
+# Copyright (c) 2019-2020, FusionAuth, All Rights Reserved
+#
+# Licensed under the Apache License, Version 2.0 (the \"License\");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+
+# ...
+
+# Webhook transaction failure
+[WebhookTransactionException]=One or more webhooks returned an invalid response or were unreachable. Based on your transaction configuration, your action cannot be completed.
+```
+
+The file is approximately 200 lines in length so the above is an excerpt. Open it in your favorite text editor.
+
+===== Modifying the Messages File
+
+You are going to add both placeholders for the text input boxes as well as custom validation messages. To add the placeholders, add values to the `Custom Registration` section. 
+
+Maintaining sections in this file isn't required since it's not a `.ini` file. However, it's a good idea to change only what is needed and not restructure the entire file. Upgrades to FusionAuth will add more properties and you will have to merge your changes in. Search for the section starting with `Custom Registration forms`.
+
+The keys of the messages file must match the field keys for the registration form.
+
+To add the placeholders for the custom and default input fields, add these lines:
+
+```properties
+# ...
+user.firstName=Your first name
+user.mobilePhone=Your mobile phone num
+registration.data.minprice=Minimum home price
+registration.data.maxprice=Maximum home price
+registration.data.geographicarea=Where are you looking?
+# ...
+```
+
+To add validation messages, search for `# Custom Registration form validation errors`. You'll add the error messages there.
+
+Each error message takes the form: `[errortype]fieldname`. Look at the `Default validation errors` section to see the list of valid `errortype`s. The field name is the keyname for the field. For example, to display a user friendly error message when required price range information is omitted or invalid, add these properties:
+
+```properties
+[invalid]registration.data.minprice=Minimum home price required
+[invalid]registration.data.maxprice=Maximum home price required
+```
+
+These messages are displayed to the user when the minimum or maximum prices are `invalid`. Because these fields have the `number` datatype, they are `invalid` any time the user input is not a number, including the empty string.
```

Comment:
See above comment.  Depending upon how much effort it is - you may want to recreate this scenario with the latest release to ensure the messages are correct.

Reply:
Awesome, thanks! 1.19.4 behaves as you outline. I updated the guide.